File: tobas_setup_assistant/src/tobas_setup_assistant/parameter_getters/spin_box.py
from typing import override
from typing import Optional
from PyQt5.QtCore import Qt, pyqtSignal, pyqtSlot

from tobas_std_tools_py.string import convert_superscript
from tobas_rqt_tools.widgets import SpinBox

from .base import ParamGetterWidget


class ParamGetterWidget_SpinBox(ParamGetterWidget[int]):
    value_changed = pyqtSignal(int)

    def __init__(
        self,
        param_name: str,
        description_text: Optional[str] = None,
        minimum: int = -(1 << 31),
        maximum: int = (1 << 31) - 1,
        single_step: int = 1,
        default: Optional[int] = None,
        suffix: str = "",
    ) -> None:
        assert minimum < maximum
        assert single_step > 0

        super().__init__(param_name, description_text)

        self._spin_box = SpinBox()
        self._rows.addWidget(self._spin_box)

        self._spin_box.setMinimum(minimum)
        self._spin_box.setMaximum(maximum)
        self._spin_box.setSingleStep(single_step)
        if default is not None:
            assert minimum <= default <= maximum
            self._spin_box.setValue(default)
        self._spin_box.setSuffix(convert_superscript(suffix))
        self._spin_box.setFocusPolicy(Qt.StrongFocus)
        self._spin_box.valueChanged.connect(self._on_value_changed)

    @override
    def get(self) -> int:
        return self._spin_box.value()

    @override
    def set(self, src: int) -> None:
        self._spin_box.setValue(src)

    @pyqtSlot(int)
    def _on_value_changed(self, value: int) -> None:
        self.value_changed.emit(value)

File: tobas_gui/tobas_setup_assistant/src/param_getters/spin_box.cpp
#include <tobas_std_tools/check.hpp>
#include <tobas_std_tools/string.hpp>

#include "tobas_setup_assistant/param_getters/spin_box.hpp"

namespace gui
{
namespace setup_assistant
{
ParamGetterWidget_SpinBox::ParamGetterWidget_SpinBox(
  const QString& param_name,
  const QString& description_text,
  int minimum,
  int maximum,
  int single_step,
  std::optional<int> _default,
  const QString& suffix)
  : super(param_name, description_text)
{
  TOBAS_CHECK(minimum <= maximum);
  TOBAS_CHECK(single_step > 0);

  spin_box_ = new qt::SpinBox();
  rows_->addWidget(spin_box_);

  spin_box_->setMinimum(minimum);
  spin_box_->setMaximum(maximum);
  spin_box_->setSingleStep(single_step);
  if (_default.has_value())
  {
    TOBAS_CHECK(minimum <= _default && _default <= maximum);
    spin_box_->setValue(_default.value());
  }
  spin_box_->setSuffix(QString::fromStdString(tobas_std::convertToSuperscript(suffix.toStdString())));
  spin_box_->setFocusPolicy(Qt::StrongFocus);

  connect(spin_box_, QOverload<int>::of(&qt::SpinBox::valueChanged), this, &self::onValueChanged);
}

int ParamGetterWidget_SpinBox::get() const
{
  return spin_box_->value();
}

bool ParamGetterWidget_SpinBox::set(const int& src)
{
  if (src < spin_box_->minimum() || spin_box_->maximum() < src)
    return false;

  spin_box_->setValue(src);
  return true;
}

void ParamGetterWidget_SpinBox::onValueChanged(int value)
{
  Q_EMIT valueChanged(value);
}
}  // namespace setup_assistant
}  // namespace gui
